### ISSUE TITLE
Buy-Side Listings

### ISSUE THREAD

**Issue by micahalcorn** [OriginProtocol/origin#2881]
Many marketplaces require the buyer to "go first". For example, a ridesharing buyer requests to be transported from point A to point B. One or more sellers are presented with that request along with a price and a timeframe (both likely generated by an intermediate dispatching service), and they can agree to provide the service.

[We previously supported "Wanted" listing categories, but these have since been removed](https://github.com/OriginProtocol/origin/issues/948).

This functionality should be introduced at the contract level or we should at least confirm that our current marketplace contract can be used to accommodate this use case.

At the DApp layer, we would need to figure out how to organize and display these listings. Currently, our concepts of "Listings" and "Sales" presume seller-initiated transactions.
-----

At Origin, we use [ZenHub](https://www.zenhub.com/) to manage our engineering tasks and product development. [Download their browser extension](https://www.zenhub.com/extension) and check out our open workspace at [github.com/originprotocol/origin#zenhub](https://github.com/originprotocol/origin#zenhub).

**Comment by nick** [comment]
I don't think this will require any contract level changes as, in order for the escrow system to work effectively, both buyer and seller must agree on terms.

For ride sharing, I think the flow would be something like this:

1. Buyer makes request to centralized server with ride request details (source, destination)
2. Server finds nearby driver, works out price, and returns drivers listing ID and price to buyer
3. Buyer makes offer on driver's listing (like how you pre-pay on Uber and Lyft) with finalization window of 1 hour after estimated drop off time
4. Driver accepts offer, picks up buyer, takes to destination
5. Buyer finalizes offer, or driver finalizes after window passes

**Comment by matthewliu** [comment]
I think this "almost works" but not quite. In the Uber world, step 3 does not have a specific driver, only the route and rate. Then Uber will ping drivers sequentially until they find a driver that accepts. There's a lot of room for error here where our server lists a specific driver for the buyer to make an offer against, and then the driver refuses. I would be quite annoyed if I requested a ride and was waiting around wondering if the driver would accept my offer indefinitely. If I had to re-request every time until a seller accepted, I wouldn't use the app.

We shouldn't necessarily optimize for Uber, but I do believe that we will encounter multiple true Buyer-First use cases.

**Comment by nick** [comment]
Yea agree there are multiple buyer first cases, but I don't see why we'd need a contract change.

Let's say we did make a contract change that allows the buyer to send funds to the marketplace and any seller to accept that offer. How can the contract tell a real seller from a fake one? Someone could easily write a script to auto-accept all offers and essentially DoS the system. In that case should we allow a whitelist of sellers? Who maintains that list?

Going back to the Uber case, I think we can solve the issue by having the seller be a broker. Maybe one day Uber and Lyft are just brokers on Origin:

1. Buyer makes request to broker's server with ride request details (source, destination)
2. Broker determines price and whether they have drivers available, and returns price to buyer
3. Buyer makes offer on broker's listing
4. Broker pings drivers sequentially until they find a driver that accepts
5. Driver picks up buyer and takes to destination
6. Buyer finalizes offer, broker distributes funds to driver

Since a broker will be needed to manage driver positions and availability anyway, this could work well

**Comment by matthewliu** [comment]
I think the broker would be sufficient. 

Are there other examples we are missing? Just want to be comprehensive and make sure we can support these (not high-priority at the moment):
- Buyer requests multiple designers submit logos
- Buyer wants quote from multiple service workers (e.g. plumber, handyman)
- Buyer looking for a rare collectible

We certainly don't need to make unnecessary contract changes until we need them, but I think we should reserve the right to if our current paradigm doesn't work.

The last thing worth mentioning is eventually we will want to support "instant" transactions. Whether that just has automated offer-accept doesn't really matter to me as long as it is perceived that way to users in the UI.

CC @joshfraser

**Comment by mikeshultz** [comment]
> How can the contract tell a real seller from a fake one? Someone could easily write a script to auto-accept all offers and essentially DoS the system.

You can say the same thing about buyers now.

I assume this requires changes, but why shouldn't the roles just be reversed for this purpose?

1) Andy wants a car.  Joe is "selling" his money for a service.  Posts a WTB listing.
2) Barbara wants to pay for this money with her service.  Makes an offer.

I suppose this might require some escrow changes, but wonder if it doesn't make sense to allow either side to post some kind of escrow.  These changes could also be made to support the sale of things like ERC20/ERC721 tokens.

Just a thought.

**Comment by nick** [comment]
The difference is that the buyer must submit funds in Eth or ERC20 in order for the offer to be valid. The seller is under no obligation to accept any given offer, no matter how many are submitted. If, however, the buyer submits a 'wanted' listing with Eth and a fake seller accepts, the buyer must now go to arbitration to retrieve their funds.

Regarding designers submitting logos and quotes from multiple sellers, I think the buyer can just post a 'wanted' listing and have sellers quote them. Then the buyer makes an actual offer on the seller they have selected.

'Instant' transactions can be supported with automated offer-accept. Now we have meta transactions, the seller can just pre-sign an accept offer tx which we can submit automatically once a buyer makes a valid offer.

**Comment by mikeshultz** [comment]
Thanks for the clarification.  Personally, I'd prefer to see contract changes rather than a centralized service, but I get that that's not something we'd want to rush into.

**Comment by nick** [comment]
Yea I don't have an issue with contract changes, I just don't see how 'instant buy' would work at the contract level where escrow and arbitration is involved.

For digital goods like ERC721, though... a new contract makes sense.

**Comment by joshfraser** [comment]
I think the current contracts are sufficient for now. We can revisit if we run into more specific use-cases that aren't possible with our current contracts.
